--- ./libxnet/xnet.c.orig	Tue Apr 13 06:12:53 2004
+++ ./libxnet/xnet.c	Fri Jun 11 01:29:18 2004
@@ -3255,17 +3255,17 @@
 	{
 		__xnet_u_setstate_const(user, TS_DATA_XFER);
 		if (call) {
-			if (call->addr.maxlen < p->conn_con.RES_length)
+			if (call->addr.maxlen && call->addr.maxlen < p->conn_con.RES_length)
 				goto tbufovflw;
-			if (call->opt.maxlen < p->conn_con.OPT_length)
+			if (call->opt.maxlen && call->opt.maxlen < p->conn_con.OPT_length)
 				goto tbufovflw;
-			if (call->udata.maxlen < user->data.len)
+			if (call->udata.maxlen && call->udata.maxlen < user->data.len)
 				goto tbufovflw;
-			if ((call->addr.len = p->conn_con.RES_length))
+			if (call->addr.maxlen && (call->addr.len = p->conn_con.RES_length))
 				memcpy(call->addr.buf, (char *) p + p->conn_con.RES_offset, call->addr.len);
-			if ((call->opt.len = p->conn_con.OPT_length))
+			if (call->opt.maxlen && (call->opt.len = p->conn_con.OPT_length))
 				memcpy(call->opt.buf, (char *) p + p->conn_con.OPT_offset, call->opt.len);
-			if ((call->udata.len = user->data.len))
+			if (call->udata.maxlen && (call->udata.len = user->data.len))
 				memcpy(call->udata.buf, user->data.buf, call->udata.len);
 		}
 		__xnet_u_reset_event(user);	/* consume the event */
